### DISCUSSION TITLE
How to handle embedded assets

### DISCUSSION THREAD

**Discussion by wwderw** [premake/premake-core#1995] [Q&A]
I have got object files for embedded assets (art, audio/video, maybe text files etc).

I know one simple test script that I created is valid when I strip out the need for the assets, everything compiles just fine.  I have attempted to put them in the files{} portion and also after I put the files into a library archive (.a file) and put them in the links{} portion and tried to compile it that way.

I don't know if I'm trying to shoehorn those files in the wrong spot or maybe one of those is the correct spot, but I'm not doing something correct in the follow through.

Appreciate any help on this.

Thanks!

**Comment by Jarod42** [comment]
- What do you expect from those files? (just appears in IDE solution, be embed as resources, ...)
- Which generators do you use?

**Reply by wwderw** [reply]
This is actually for assets for a game, sprite sheets, SFX etc.  So it is supposed to be for the final product, not just what and IDE would see.

**Reply by Jarod42** [reply]
What do you expect from **build system**? (How do you access them from program?)
Those kinds of ressources might just be copied from installer along the executable.
Some tools might allows to embed them in specific files (even in exe file, even if it is no longer the preferred way).

**Reply by wwderw** [reply]
My full process right now is:

1. Use ld to generate the object files
2.  In code I will extern DATA and SIZE for each resource that is now an o file(s)
3.  Typically whatever lib I'm use (be it some type of server lib or in this cause a game lib) will have a function of some derivation of ImageFromeMemory(), WaveFromMemory() etc that allows me to use those values that I got in number 2 above and I'll use that to access those files in the program.
4.  When I go to build, I'll use g++ main.cpp *.o -L [whatever libs I'm linking to] -o finalBinary

My goal is to have them linked/compiled into the final binary (as long as the HD footprint isn't outrageous for one file, which mine typically are not).

**Comment by Jarod42** [comment]
Not familiar with `ld` to do that. `winres`, .qrc (for Qt), [xxd](https://linux.die.net/man/1/xxd) to create C-array as resources seems more natural for me.

Normally, in premake, non-source file code in `files {..}` are simply ignored from build and only shown in IDE.

Custom build for specific file (so `filter` usage) (with [buildcommands](https://premake.github.io/docs/buildcommands)) or [Custom rules](https://premake.github.io/docs/Custom-Rules) should be possible to do what you want.

Sharing working command lines might help to see what you are doing.

**Comment by samthesuperhero** [comment]
Hi @Jarod42,

for me still actual,

here is the discussion,
[How do I automatically include folders and files to the deployment project in visual studio?](https://stackoverflow.com/questions/63332425/how-do-i-automatically-include-folders-and-files-to-the-deployment-project-in-vi)

 - any ideas what to specify in `premake5.lua` script so that Visual Studio solution / project is configured to copy contents / assets files to specified bin folder(s) ?

**Reply by Jarod42** [reply]
@samthesuperhero : you seems to want to copy files in targetdir (whereas OP want to embed them), so, in your case, something along:

```
postbuildcommands {
  "{COPYDIR} your_dir %[%{!cfg.targetdir}]",           -- for whole directory
  "{COPY} your_file %[%{!cfg.targetdir}/your_file]", -- for specific file
}
```
See [Tokens](https://premake.github.io/docs/Tokens/) and [`postbuildcommands`](https://premake.github.io/docs/postbuildcommands) for more details.

**Reply by samthesuperhero** [reply]
Strange this is what I get:
```
Severity	Code	Description	Project	File	Line	Suppression State
Error	MSB3073	The command "xcopy /Q /E /Y /I assets "E:\dev\raylib\rlImGui-template\bin\Debug"
:VCEnd" exited with code 4.	rlimgui-template	D:\MicrosoftVisualStudio\MSBuild\Microsoft\VC\v170\Microsoft.CppCommon.targets	160	
```

**Reply by Jarod42** [reply]
Probably`%[your_dir_relative_to_premake_script]` or `your_dir_relative_to_project`.

**Reply by samthesuperhero** [reply]
Works! Thank you!!
